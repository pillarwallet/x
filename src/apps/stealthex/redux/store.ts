import { configureStore, TypedStartListening } from '@reduxjs/toolkit';
import exchangeReducer from './reducers/exchange';
import { exchange } from '../lib/backend/api';

export const store = configureStore({
    reducer: {
        exchange: exchangeReducer,
        [exchange.reducerPath]: exchange.reducer
    },
    middleware: getDefaultMiddleware =>
        getDefaultMiddleware({ serializableCheck: false }).concat(exchange.middleware),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppStartListening = TypedStartListening<RootState, AppDispatch>; 