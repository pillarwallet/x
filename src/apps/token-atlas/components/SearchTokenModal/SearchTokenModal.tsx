// hooks
import { useAppDispatch, useAppSelector } from '../../hooks/useReducerHooks';

// reducer
import {
  setIsAllChainsVisible,
  setIsSearchTokenModalOpen,
  setIsSelectChainDropdownOpen,
  setSearchTokenResult,
  setSelectedChain,
} from '../../reducer/tokenAtlasSlice';

// utils
import { CompatibleChains } from '../../../../utils/blockchain';

// images
import CloseCircle from '../../images/close-circle.svg';

// components
import SelectChainDropdown from '../SelectChainDropdown/SelectChainDropdown';
import TokensSearchInput from '../TokensSearchInput/TokensSearchInput';
import TokensSearchResult from '../TokensSearchResult/TokensSearchResult';
import TokensSlider from '../TokensSlider/TokensSlider';

const SearchTokenModal = () => {
  const dispatch = useAppDispatch();
  const isSearchTokenModalOpen = useAppSelector(
    (state) => state.tokenAtlas.isSearchTokenModalOpen as boolean
  );
  const isSelectChainDropdownOpen = useAppSelector(
    (state) => state.tokenAtlas.isSelectChainDropdownOpen as boolean
  );

  // select all chainsId of tokens available in the list for swap token
  const uniqueChains = CompatibleChains.map((chain) => chain.chainId);

  const handleOnCloseSearchModal = () => {
    dispatch(setIsSearchTokenModalOpen(false));
    dispatch(setSelectedChain({ chainId: 0, chainName: 'all' }));
    dispatch(setSearchTokenResult(undefined));
    dispatch(setIsAllChainsVisible(false));
  };

  if (!isSearchTokenModalOpen) return null;

  return (
    <>
      <div
        id="token-atlas-search-modal"
        className="fixed inset-0 bg-[#222222]/[.4] backdrop-blur-sm z-30"
      >
        <div className="fixed inset-0 flex items-center justify-center z-40 mobile:items-start tablet:items-start mt-20">
          <div className="relative flex flex-col w-full desktop:max-w-[600px] tablet:max-w-[600px] mobile:max-w-full mobile:mx-4 bg-dark_grey border-[1px] border-medium_grey rounded-2xl overflow-y-auto max-h-[75vh] mb-20">
            <div className="flex gap-4 px-8 pt-8 pb-6 border-b-4 border-b-medium_grey">
              <TokensSearchInput
                className={`${
                  !isSelectChainDropdownOpen ? 'basis-3/5' : 'max-w-14'
                }`}
                onClick={() =>
                  isSelectChainDropdownOpen &&
                  dispatch(setIsSelectChainDropdownOpen(false))
                }
              />
              <SelectChainDropdown
                className={`${
                  !isSelectChainDropdownOpen ? 'basis-2/5' : 'w-full'
                }`}
                options={uniqueChains}
              />
            </div>
            <div className="flex px-8 py-4 border-b-[1px] border-b-medium_grey">
              <TokensSlider />
            </div>
            <div className="flex px-8 py-4">
              <TokensSearchResult />
            </div>
          </div>
        </div>
      </div>

      <button
        type="button"
        id="token-atlas-close-search-modal"
        className="fixed top-0 right-0 w-[50px] h-[50px] mt-10 mr-[60px] mb-20 mobile:mt-4 mobile:mr-4 z-50"
        onClick={() => handleOnCloseSearchModal()}
      >
        <img src={CloseCircle} alt="Close" />
      </button>
    </>
  );
};

export default SearchTokenModal;
