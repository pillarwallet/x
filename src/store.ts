/* eslint-disable @typescript-eslint/no-use-before-define */
// Core
import {
  Middleware,
  Reducer,
  combineReducers,
  configureStore,
  createDynamicMiddleware,
} from '@reduxjs/toolkit';
import { setupListeners } from '@reduxjs/toolkit/query';

// Services
import swapSlice from './apps/the-exchange/reducer/theExchangeSlice';
import tokenAtlasSlice from './apps/token-atlas/reducer/tokenAtlasSlice';
import { pillarXApiPresence } from './services/pillarXApiPresence';
import { pillarXApiTransactionsHistory } from './services/pillarXApiTransactionsHistory';
import { pillarXApiWaitlist } from './services/pillarXApiWaitlist';

// Initialisation
const dynamicMiddleware = createDynamicMiddleware();
const middlewareReducers: { [key: string]: Reducer } = {};

/**
 * @name addReducer
 * @description addReducer is a function that allows anyone
 * working with PillarX to add reducers to the store
 * from their own application.
 *
 * @param newReducer
 */
export const addReducer = (newReducer: {
  reducerPath: string;
  reducer: Reducer;
}) => {
  middlewareReducers[newReducer.reducerPath as string] = newReducer.reducer;
  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * @name addMiddleware
 * @description addMiddleware is a function that allows anyone
 * working with PillarX to add RTK middleware to the store
 * from their own application.
 *
 * @param newMiddleware
 */
export const addMiddleware = (newMiddleware: {
  reducerPath: string;
  reducer: Reducer;
  middleware: Middleware;
}) => {
  middlewareReducers[newMiddleware.reducerPath as string] =
    newMiddleware.reducer;
  dynamicMiddleware.addMiddleware(newMiddleware.middleware);

  store.replaceReducer(combineReducers(middlewareReducers));
};

/**
 * Export a store from RTK
 */
export const store = configureStore({
  // Empty reducer for now - the addMiddleware function
  // below will dynamically regenerate the reducers required
  // from the middleware functions.
  reducer: {},
  // Adding the api middleware enables caching, invalidation, polling,
  // and other useful features of `rtk-query`.
  // Note: here we have added dynamicMiddleware.middleware
  // which instructs the store to use the middleware that we
  // will add later.
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().prepend(dynamicMiddleware.middleware),
});

/**
 * Whilst we're here we'll immediately mount the
 * PillarXApi middleware. This will contain all the
 * core API functionality for the PillarX application.
 */
addMiddleware(pillarXApiWaitlist);
addMiddleware(pillarXApiPresence);
addMiddleware(pillarXApiTransactionsHistory);
addReducer(swapSlice);
addReducer(tokenAtlasSlice);

// optional, but required for refetchOnFocus/refetchOnReconnect behaviors
// see `setupListeners` docs - takes an optional callback as the 2nd arg for customization
setupListeners(store.dispatch);

// // Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
