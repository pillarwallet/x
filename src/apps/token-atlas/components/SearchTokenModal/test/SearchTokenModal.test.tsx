import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom'; // Import MemoryRouter
import renderer from 'react-test-renderer';

// redux store
import { Provider } from 'react-redux';
import { store } from '../../../../../store';

// reducer
import {
  setIsAllChainsVisible,
  setIsSearchTokenModalOpen,
  setIsSelectChainDropdownOpen,
  setSearchTokenResult,
  setSelectedChain,
} from '../../../reducer/tokenAtlasSlice';

// components
import SearchTokenModal from '../SearchTokenModal';

describe('<SearchTokenModal />', () => {
  beforeEach(() => {
    store.dispatch(setSearchTokenResult(undefined));
    store.dispatch(setIsAllChainsVisible(false));
    store.dispatch(setSelectedChain({ chainId: 0, chainName: 'all' }));
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <SearchTokenModal />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when modal is open', () => {
    store.dispatch(setIsSearchTokenModalOpen(true));

    render(
      <Provider store={store}>
        <MemoryRouter>
          <SearchTokenModal />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByText('Start searching for tokens.')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /close/i })).toBeInTheDocument();
  });

  it('renders nothing when modal is closed', () => {
    store.dispatch(setIsSearchTokenModalOpen(false));

    render(
      <Provider store={store}>
        <MemoryRouter>
          <SearchTokenModal />
        </MemoryRouter>
      </Provider>
    );

    expect(
      screen.queryByText('Start searching for tokens.')
    ).not.toBeInTheDocument();
    expect(
      screen.queryByRole('button', { name: /close/i })
    ).not.toBeInTheDocument();
  });

  it('closes the modal when the close button is clicked', () => {
    store.dispatch(setIsSearchTokenModalOpen(true));
    render(
      <Provider store={store}>
        <MemoryRouter>
          <SearchTokenModal />
        </MemoryRouter>
      </Provider>
    );

    fireEvent.click(screen.getByRole('button', { name: /close/i }));

    expect(store.getState().tokenAtlas.isSearchTokenModalOpen).toBe(false);
    expect(store.getState().tokenAtlas.selectedChain).toEqual({
      chainId: 0,
      chainName: 'all',
    });
    expect(store.getState().tokenAtlas.searchTokenResult).toEqual(undefined);
    expect(store.getState().tokenAtlas.isAllChainsVisible).toBe(false);
  });

  it('handles SelectChainDropdown visibility correctly when opened', () => {
    store.dispatch(setIsSearchTokenModalOpen(true));
    store.dispatch(setIsSelectChainDropdownOpen(true));
    render(
      <Provider store={store}>
        <MemoryRouter>
          <SearchTokenModal />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByTestId('select-chain-dropdown')).toHaveClass('w-full');
  });

  it('handles SelectChainDropdown visibility correctly when closed', () => {
    store.dispatch(setIsSearchTokenModalOpen(true));
    store.dispatch(setIsSelectChainDropdownOpen(false));

    render(
      <Provider store={store}>
        <MemoryRouter>
          <SearchTokenModal />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByTestId('select-chain-dropdown')).toHaveClass(
      'basis-2/5'
    );
  });
});
