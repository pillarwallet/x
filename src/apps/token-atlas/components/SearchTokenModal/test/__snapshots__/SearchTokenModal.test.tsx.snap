// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<SearchTokenModal /> > renders correctly and matches snapshot 1`] = `null`;

exports[`<SearchTokenModal /> renders correctly and matches snapshot 1`] = `null`;
